src/Reflector/Generic.php

```diff
@@ -0,0 +1,196 @@
+                $symbol->getName()
+            ));
+        }
+
+        $locatedSource = $this->sourceLocator->__invoke($symbol);
+        $reflection = $this->reflectFromLocatedSource($symbol, $locatedSource);
+
+        return $reflection;
+    }
+
+    /**
+     * Given an array of Reflections, try to find the symbol
+     *
+     * @param Reflection[] $reflections
+     * @param Symbol $symbol
+     * @return Reflection
+     */
+    private function findInArray($reflections, Symbol $symbol)
+    {
+        foreach ($reflections as $reflection) {
+            if ($reflection->getName() == $symbol->getName()) {
+                return $reflection;
+            }
+        }
+
+        throw new \UnexpectedValueException(sprintf(
+            '%s "%s" could not be found to load',
+            $symbol->getDisplayType(),
+            $symbol->getName()
+        ));
+    }
+
+    /**
+     * Read all the symbols from a LocatedSource and find the specified symbol
+     *
+     * @param Symbol $symbol
+     * @param LocatedSource $locatedSource
+     * @return Reflection
+     */
+    private function reflectFromLocatedSource(
+        Symbol $symbol,
+        LocatedSource $locatedSource
+    ) {
+        $reflections = $this->getReflections($locatedSource, $symbol);
+        return $this->findInArray($reflections, $symbol);
+    }
+
+    /**
+     * @param Node $node
+     * @return Reflection|null
+     */
+    private function reflectNode(Node $node, Node\Stmt\Namespace_ $namespace = null, $filename = null)
+    {
+        if ($node instanceof Node\Stmt\Class_) {
+            return ReflectionClass::createFromNode(
+                $node,
+                $namespace,
+                $filename
+            );
+        }
+
+        return null;
+    }
+
+    /**
+     * Process and reflect all the matching symbols found inside a namespace node
+     *
+     * @param Node\Stmt\Namespace_ $namespace
+     * @param Symbol $symbol
+     * @param string|null $filename
+     * @return Reflection[]
+     */
+    private function reflectFromNamespace(
+        Node\Stmt\Namespace_ $namespace,
+        Symbol $symbol,
+        $filename
+    ) {
+        $reflections = [];
+        foreach ($namespace->stmts as $node) {
+            $reflection = $this->reflectNode($node, $namespace, $filename);
+
+            if (null !== $reflection && $symbol->isMatchingReflector($reflection)) {
+                $reflections[] = $reflection;
+            }
+
+        }
+        return $reflections;
+    }
+
+    /**
+     * Reflect symbols from an AST. If a namespace is found, also load all the
+     * matching symbols found in the namespace
+     *
+     * @param Node[] $ast
+     * @param string|null $filename
+     * @param Symbol $symbol
+     * @return Reflection[]
+     */
+    private function reflectFromTree(array $ast, $filename, Symbol $symbol)
+    {
+        $reflections = [];
+        foreach ($ast as $node) {
+            switch (get_class($node)) {
```

Comment:
This is a dangerous way of switching between node types, especially considering that they are not `final`.